auth-jwt-service/src/main/java/io/stargate/auth/jwt/AuthzJwtService.java

```diff
@@ -0,0 +1,186 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.auth.jwt;
+
+import static io.stargate.auth.jwt.AuthnJwtService.CLAIMS_FIELD;
+import static io.stargate.auth.jwt.AuthnJwtService.STARGATE_PREFIX;
+
+import io.stargate.auth.AuthorizationService;
+import io.stargate.auth.UnauthorizedException;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.schema.Column;
+import io.stargate.db.schema.Table;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.regex.Pattern;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AuthzJwtService implements AuthorizationService {
+
+  private static final Logger log = LoggerFactory.getLogger(AuthzJwtService.class);
+  private final Pattern tokenPattern = Pattern.compile("\\.");
+
+  /**
+   * Using the provided JWT and the claims it contains will perform pre-authorization where
+   * possible, executes the query provided, and then authorizes the response of the query.
+   *
+   * @param action A {@link QueryBuilder} object to be executed and authorized against a JWT.
+   * @param token The authenticated JWT token to use for authorization.
+   * @param primaryKeyValues A list of primary key values that will be used in the query and should
+   *     be authorized against the JWT.
+   * @param tableMetadata The {@link Table} that will be queried against.
+   * @return On success will return the result of the query and otherwise will return an exception
+   *     relating to the failure to authorize.
+   * @throws Exception An exception relating to the failure to authorize.
+   */
+  @Override
+  public ResultSet authorizedDataRead(
+      Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
+      throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckDataReadWrite(stargateClaims, primaryKeyValues, tableMetadata);
+
+    ResultSet result = action.call();
+
+    if (result == null) {
+      return null;
+    }
+
+    return result.withRowInspector(
+        row -> {
+          if (row == null) {
+            return true;
+          }
+
+          for (Column col : row.columns()) {
+            if (stargateClaims.has(STARGATE_PREFIX + col.name())) {
+
+              String stargateClaimValue;
+              try {
+                stargateClaimValue = stargateClaims.getString(STARGATE_PREFIX + col.name());
+              } catch (JSONException e) {
+                log.warn("Unable to get stargate claim for " + STARGATE_PREFIX + col.name());
+                return false;
+              }
+              String columnValue = row.getString(col.name());
```

Comment:
Sorry, I'm not fully understanding this. Is this the arbitrary values of columns from the DB? Why is it comparing column values in the DB to values in the JWT payload?

Reply:
Yes. The idea is that this allows you to do RLAC with JWT claims.

So let's say you have a table `shopping_cart` with pk of `userid` and you want to write a JS app that allows users to login and see the contents of their cart. Rather than having a bunch of roles or doing some sort of backend filtering instead you can assign a new claim to the user's JWT with their userid. Then when you JS app makes a request to graphql with the JWT stargate will return only the data that the user is allowed to see. 

By looking through the claims and comparing to columns all of this can be done completely on the fly by adjusting what claims are in the JWTs rather than making config and/or code changes to filter different columns or tables.

Reply:
Thanks for the explanation. Cool idea.